humility-core/src/arch/mod.rs

```diff
@@ -0,0 +1,352 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::hubris::{HubrisArchive, HubrisStruct};
+use crate::regs;
+use crate::regs::arm::ARMRegister;
+use crate::regs::rv::RVRegister;
+use crate::regs::Register;
+use anyhow::{bail, Result};
+use capstone::arch;
+use capstone::arch::ArchOperand;
+use capstone::prelude::*;
+use num_traits::cast::ToPrimitive;
+use num_traits::FromPrimitive;
+use std::collections::BTreeMap;
+use strum::IntoEnumIterator;
+
+pub mod arm;
+pub mod rv;
+
+#[derive(Debug)]
+pub struct Arch {
+    arch: Option<u16>,
+    abi_size: Option<u8>,
+}
+
+impl Default for Arch {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Arch {
```

Comment:
I like your idea of having a separate Arch and then having methods that provide the required arch specific details. Just one thing I noted is that you have checks in all the methods because you need to ensure that the arch is supplied correctly. However, we really just need to check that once. Something like this could probably simply the structure:

```

struct ARMArch {}

impl ARMArch {
    fn new(_abi_size: u8) {
        ARMArch {}
    }
}

impl Arch for ARMArch {
    fn get_syscall_insn(&self) -> u32 {
        arch::riscv::RiscVInsn::RISCV_INS_ECALL as u32
    }
    
    fn get_ret_reg(&self) -> Register {
        Register::Arm(ARMRegister::LR)
    }
    ...
}

pub struct RVArch {
    pub abi_size: u8,
}

impl RVArch {
    fn new(abi_size: u8) {
        RVArch { abisize }
    }
}

impl Arch for RVArch {
    fn get_syscall_insn(&self) -> u32 {
        arch::arm::ArmInsn::ARM_INS_SVC as u32
    }
    fn get_ret_reg(&self) -> Register {
        Register::RiscV(RVRegister::RA)
    }
    ...
}

pub trait Arch {
    fn get_syscall_insn(&self) -> u32;
    fn get_ret_reg(&self) -> Register;
    ...
}

pub fn get_arch(arch: u16, abi_size: u8) -> Box<dyn Arch> {
    if arch == goblin::elf::header::EM_ARM {
        Box::new(ARMArch::new(abi_size))
    } else if arch == goblin::elf::header::EM_RISCV {
        Box::new(RVArch::new(abi_size))
    } else {
        unimplemented!();
    }
}
```
WIth this you can eliminate the arch check in each function and also move the RVArch and ARMArch structures to live within their respective arch files i.e. arm.rs and rv.rs) What do you think?

Reply:
Playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c41efb3c9f7b5c389700ad176d8dd2da

Reply:
Another playground [link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=57bbe1c71b57e985550024a47a4763d3) that implements a method to return Reg Trait from Arch Trait implementation.

Reply:
I went ahead and did a trait for `Arch`. 

When it comes to the registers though, I think using a `Box<dyn Register>` is more effort than it is worth.  Lots of Humility relies on traits implemented on `Register`, and these traits are not implemented on `dyn Register`.   I would have to implement these, rather than use the default implementations, as well as the boilerplate to box everything up.  I will think about it, but will probably leave it out for know unless you think it is a lot better.